st2api/st2api/controllers/v1/actionexecutions.py

```diff
@@ -365,6 +366,61 @@ def get(self, id, attribute, requester_user):
         return Response(json=result, status=http_client.OK)
 
 
+class ActionExecutionRawResultController(BaseActionExecutionNestedController):
+    def get(self, id, requester_user, download=False):
+        """
+        Retrieve raw action execution result object as a string or optionally force result download
+        as a file.
+
+        This is primarily to be used in scenarios where executions contain large results and JSON
+        loading and parsing it can be slow (e.g. in the st2web) and we just want to display raw
+        result.
+
+        Handles requests:
+
+            GET /executions/<id>/result/raw[?download=1]
+
+        :rtype: ``str``
+        """
+        download = str(download).lower() in ["1", "true"]
+
+        # NOTE: Here we intentionally use as_pymongo() to avoid mongoengine layer even for old style
+        # data
+        try:
+            result = (
+                self.access.impl.model.objects.filter(id=id)
+                .only("result")
+                .as_pymongo()[0]
+            )
+        except IndexError:
+            # TODO: return 404
+            return ""
+
+        if isinstance(result["result"], dict):
+            # For backward compatibility we also support old non JSON field storage format
+            string_data = orjson.dumps(result["result"], option=orjson.OPT_INDENT_2)
+        else:
+            # For new JSON storage format we just use raw value since it's already JSON serialized
+            # string
+            string_data = result["result"]
+            # TODO: Should we re-serialize for pretty print purposes. It adds around 80ms overhead
+            # for large executions so it's not terrible - e.g. 30ms for using raw results vs 100ms
+            # for reserializing it with pretty indent
+            # string_data = orjson.dumps(orjson.loads(result["result"]), option=orjson.OPT_INDENT_2)
```

Comment:
Maybe add another query param, `pretty=1` (default 0) so people can opt into pretty formatting without copying it into some other service.

Reply:
Yeah, that's what I ended up doing in the end (``pretty_format`` query parameter).